cpp/src/mlt/decode/geometry.hpp

```diff
@@ -0,0 +1,430 @@
+#pragma once
+
+#include <mlt/decode/int.hpp>
+#include <mlt/feature.hpp>
+#include <mlt/geometry.hpp>
+#include <mlt/metadata/stream.hpp>
+#include <mlt/metadata/tileset.hpp>
+#include <mlt/util/buffer_stream.hpp>
+#include <mlt/util/noncopyable.hpp>
+
+#include <algorithm>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace mlt::geometry {
+struct GeometryColumn : public util::noncopyable {
+    GeometryColumn() = default;
+    ~GeometryColumn() noexcept = default;
+
+    GeometryColumn(GeometryColumn&&) noexcept = default;
+    GeometryColumn& operator=(GeometryColumn&&) = delete;
+
+    std::vector<metadata::tileset::GeometryType> geometryTypes;
+    std::vector<std::uint32_t> geometryOffsets;
+    std::vector<std::uint32_t> partOffsets;
+    std::vector<std::uint32_t> ringOffsets;
+    std::vector<std::uint32_t> vertexOffsets;
+    std::vector<std::int32_t> vertices;
+};
+} // namespace mlt::geometry
+
+namespace mlt::decoder {
+
+class GeometryDecoder {
+    GeometryFactory geometryFactory;
+
+public:
+    using GeometryColumn = geometry::GeometryColumn;
+
+    GeometryColumn decodeGeometryColumn(BufferStream& tileData,
+                                        const metadata::tileset::Column& column,
+                                        std::uint32_t numStreams) noexcept(false) {
+        using namespace util::decoding;
+        using namespace metadata::stream;
+        using namespace metadata::tileset;
+
+        GeometryColumn geomColumn;
+
+        const auto geomTypeMetadata = StreamMetadata::decode(tileData);
+        intDecoder.decodeIntStream<std::uint32_t, std::uint32_t, GeometryType>(
+            tileData, geomColumn.geometryTypes, *geomTypeMetadata);
+
+        for (std::uint32_t i = 1; i < numStreams; ++i) {
+            const auto geomStreamMetadata = StreamMetadata::decode(tileData);
+            switch (geomStreamMetadata->getPhysicalStreamType()) {
+                case PhysicalStreamType::LENGTH: {
+                    if (!geomStreamMetadata->getLogicalStreamType() ||
+                        !geomStreamMetadata->getLogicalStreamType()->getLengthType()) {
+                        throw std::runtime_error("Length stream missing logical type: " + column.name);
```

Comment:
Instead of throwing exceptions you could consider using `std::expected`. Makes it a bit easier to test as well as to handle specific errors.